### ISSUE TITLE
Memory leak ?

### ISSUE THREAD

**Issue by gandroz** [TensorSpeech/TensorFlowASR#136]
With the new config, I got a memory leak which ends by a killed process

![image](https://user-images.githubusercontent.com/38106881/106523317-eee4ef00-64ae-11eb-8ce4-3b315570d8dc.png)

```
learning_config:
  train_dataset_config:
    use_tf: True
    stage: train
    augmentation_config:
      after:
        time_masking:
          num_masks: 10
          mask_factor: 100
          p_upperbound: 0.05
        freq_masking:
          num_masks: 1
          mask_factor: 27
    data_paths:
      - /data/datasets/LibriSpeech/train-clean-100/transcripts.tsv
      - /data/datasets/LibriSpeech/train-clean-360/transcripts.tsv
      - /data/datasets/LibriSpeech/train-other-500/transcripts.tsv
    tfrecords_dir: /data/datasets/Speech/LibriSpeech/tfrecords-test
    shuffle: True
    cache: True
    buffer_size: 100
    drop_remainder: True

  eval_dataset_config:
    use_tf: True
    stage: eval
    data_paths:
      - /data/datasets/LibriSpeech/dev-clean/transcripts.tsv
      - /data/datasets/LibriSpeech/dev-other/transcripts.tsv
    tfrecords_dir: /data/datasets/Speech/LibriSpeech/tfrecords-test
    shuffle: False
    cache: True
    buffer_size: 100
    drop_remainder: True
```

The RAM starts at 28GB then increases to 30GB plus 30GB of swap

As fare as I understand it, when the training dataset is created, all files of all provided datasets (`train-clean-100`, `train-clean-360` ...) are read, preprocessed (fbank are constructed here), then the dataset is cached, shuffled, padded per batch and prefetched. As the whole dataset is cached at the early beginning i.e. at the creation time, I cant see why and when RAM keeps increasing. It is as if the dataset is put into cache twice ?

**Comment by usimarit** [comment]
@gandroz As far as I understand, caching the whole dataset will load the whole dataset in memory, if the memory is not large enough it will cause OOM.
Did you use ASRSliceDataset or ASRTFRecordDataset?

**Comment by gandroz** [comment]
Oh ok, the whole dataset is loaded... with more than 600GB that's not surprising it does not fit in RAM lol
And I use ASRTFRecordDataset, I hoped training speed would increase, but not sadly no
